### DISCUSSION TITLE
How to access ServerResponse object in GraphqlContext

### DISCUSSION THREAD

**Discussion by asmi24886** [ExpediaGroup/graphql-kotlin#1095] [Q&A]
Version 3.x.x supported access to server response instance when implementing the graphql context. 
I have to set a header in the response while graphql sends the response back. 
How to do it in 4.x.x given the following code - 

```kotlin
data class CustomContext(val value: String) : GraphQLContext

class CustomFactory : GraphQLContextFactory<CustomContext, ServerRequest> {
    suspend fun generateContext(request: Request): Context? {
       //Some code to return context
    }
}
```

In 3.x.x it was much more convenient - 

```kotlin
@Component
class MyGraphQLContextFactory: GraphQLContextFactory<MyGraphQLContext> {
    override suspend fun generateContext(
        request: ServerHttpRequest,
        response: ServerHttpResponse
    ): MyGraphQLContext = MyGraphQLContext(
        myCustomValue = request.headers.getFirst("MyHeader") ?: "defaultValue"
    )
}
```

**Comment by smyrick** [comment]
Hmm, yes it does seem we removed the response from the context factory. This could easily be added back in. As of right now though it would have to be accomplished with the `GraphQLServer` object

**Reply by smyrick** [reply]
I am wondering though if mutating the response in the `ContextFactory` really is the best pattern. Should we instead force the context to only handle the context generation and the mutating of the HTTP request and response is handled in the server specific code?

**Reply by dariuszkuc** [reply]
I'm wondering what is the use case for mutating the response in the context factory (vs a webfilter).

**Reply by asmi24886** [reply]
The use case is we are using graphql as an aggregator. We call a service that sends a token, and we need to inject that token in the response header. 
The problem with passing it in the response body is that the consumer might not include it in the query response fields and thus not get it.

**Reply by asmi24886** [reply]
Please note that we are not at all modifying the response in the context factory. We pass the request and response to the query resolver function via context. We get the response from the context when after data arrives from external source. We set response header from fetching a token from the external web service response. 
I was wondering if the original DataFetchingEnvironment object can be made accessible in the resolvers. That would also be a convenient way. 
But I do believe keeping the Server Response object accessible via the context factory is an extremely convenient way.

**Reply by smyrick** [reply]
The `DataFetchingEnvironment` is available in the resolvers. If you add it as an argument it will be injected, just like the context.

https://expediagroup.github.io/graphql-kotlin/docs/schema-generator/execution/data-fetching-environment

**Reply by asmi24886** [reply]
Hi did get DataFetchingEnvironment as suggested. Please tell me what would be the most convenient way to access ServerResponse inside my query resolver method? 
In need to set a response header from a data.. that comes after calling an external webservice from inside the resolver method. Please help me out with this and suggest/implement a convenient way to access response.  
Implementing subscription context will require changes. The main idea to use this library was tor its convenience. Please make the response available. That will be great help @smyrick @dariuszkuc

**Reply by smyrick** [reply]
Even in the `3.x.x` the [GraphQLContextFactory](https://github.com/ExpediaGroup/graphql-kotlin/blob/3.x.x/graphql-kotlin-spring-server/src/main/kotlin/com/expediagroup/graphql/spring/execution/GraphQLContextFactory.kt) doesn't give you access to the `data` response. This means you are calling some service today in your context factory and modifying the `ServerHttpResponse`.

You can still do exactly that by writing your own `WebFilter` in Spring.


```kotlin
@Component
public class AddResponseHeaderWebFilter implements WebFilter {

    @Override
    public Mono<Void> filter(ServerWebExchange exchange, WebFilterChain chain) {
        exchange.getResponse()
          .getHeaders()
          .add("Example-Filter-Header",  getValueFromSerivce());
        return chain.filter(exchange);
    }
}
```

**Comment by dariuszkuc** [comment]
In both 3.x and 4.x `spring-server` you should be modifying your response headers in a webfilter. 

If it is some calculated value from incoming request that should also be a part of your GraphQL context, you could calculate it in a webfilter and store it in a subscriber context so you can then retrieve it from context in your context factory.

**Reply by smyrick** [reply]
I think this is a much better approach and clarifies what is modifying the HTTP objects vs GraphQL execution

**Reply by asmi24886** [reply]
My graphql service calls another webservice that sends a token and other relevant data, say data A.
I need to send back data A as response of graphql query and inject the token in the response header. 
With 3.x.x it is extremely convenient. 
It would be great if 4.x.x can have an easier, more convenient way to make server response accessible.

**Reply by mskonovalov** [reply]
So for those who are looking for a solution:
I had exactly the same issue: migrating from 3.x.x to 5.x.x and needed to add a token to the Header. The issue is that the token gets created inside graphql controller logic. So even if you add it to the context and then try to access it from WeFilter HTTP response will be already committed and immutable.
So how I was able to resolve that issue:

```
WebFilter { exchange, chain ->
            exchange.response.beforeCommit {
                mono {
                    setAuthHeader(ctx, exchange)
                }
            }
            chain.filter(exchange)
        }
```
So we register a hook that will be executed just before committing the HTTP response.
Although it is a bit ugly but allows to migrate to newer version with the proper fix in the future

**Comment by wickedev** [comment]
5.x.x can access the response as below.

```kotlin
class CustomGraphQLContextFactory : GraphQLContextFactory<Nothing, ServerRequest> {
    @Deprecated("The generic context object is deprecated in favor of the context map")
    override suspend fun generateContext(request: ServerRequest): Nothing? = null

    override suspend fun generateContextMap(request: ServerRequest): Map<*, Any> {
        return mapOf("request" to request)
    }
}

class SampleQuery : Query {
    fun test(env: DataFetchingEnvironment): Int {
        val response = env.graphQlContext.get<ServerRequest>("request").exchange().response
        return 0
    }
}
```

**Reply by mskonovalov** [reply]
the API for the GraphQL factory in expedia 3.x.x is 
```
override suspend fun generateContext(
        request: ServerHttpRequest,
        response: ServerHttpResponse,
    )
```
so it is `ServerHttpRequest`, not `ServerRequest`. And `ServerHttpRequest` doesn't have `exchange` as a field

But anyway thanks for the suggestion, it can be OK after upgrade
